src/bastion/src/actor/mailbox.rs

```diff
@@ -2,148 +2,199 @@ use crate::actor::actor_ref::ActorRef;
     tx: Sender<Envelope<T>>,
+    /// A field for checks that the message has been delivered to
+    /// the specific actor.
     scheduled: Arc<AtomicBool>,
 }
 
-unsafe impl<T> Send for MailboxTx<T> where T: TypedMessage {}
-unsafe impl<T> Sync for MailboxTx<T> where T: TypedMessage {}
-
 impl<T> MailboxTx<T>
 where
     T: TypedMessage,
 {
+    /// Return a new instance of MailboxTx that indicates sender.
+    pub(crate) fn new(tx: Sender<Envelope<T>>) -> Self {
+        let scheduled = Arc::new(AtomicBool::new(false));
+        MailboxTx { tx, scheduled }
+    }
+
+    /// Send the message to the actor by the channel.
     pub fn try_send(&self, msg: Envelope<T>) -> Result<()> {
         self.tx
             .try_send(msg)
             .map_err(|e| BError::ChanSend(e.to_string()))
     }
 }
 
+/// A struct that holds everything related to messages that can be
+/// retrieved from other actors. Each actor holds two queues: one for
+/// messages that come from user-defined actors, and another for
+/// internal messaging that must be handled separately.
+///
+/// For each used queue, mailbox always holds the latest requested message
+/// by a user, to guarantee that the message won't be lost if something
+/// happens wrong.
 #[derive(Clone)]
 pub struct Mailbox<T>
 where
     T: TypedMessage,
 {
-    inner: Arc<MailboxInner<T>>,
+    /// User guardian sender
+    user_tx: MailboxTx<T>,
+    /// User guardian receiver
+    user_rx: Receiver<Envelope<T>>,
+    /// System guardian receiver
+    system_rx: Receiver<Envelope<T>>,
+    /// The current processing message, received from the
+    /// latest call to the user's queue
+    latest_user_message: Option<Envelope<T>>,
+    /// The current processing message, received from the
+    /// latest call to the system's queue
+    latest_system_message: Option<Envelope<T>>,
+    /// Mailbox state machine
+    state: Arc<AtomicBox<MailboxState>>,
 }
 
+// TODO: Add calls with recv with timeout
 impl<T> Mailbox<T>
 where
     T: TypedMessage,
 {
     /// Creates a new mailbox for the actor.
-    pub fn new(inner: Arc<MailboxInner<T>>) -> Self {
-        Mailbox { inner }
+    pub(crate) fn new(system_rx: Receiver<Envelope<T>>) -> Self {
+        let (tx, user_rx) = unbounded();
+        let user_tx = MailboxTx::new(tx);
+        let state = Arc::new(AtomicBox::new(MailboxState::Scheduled));
+        let latest_user_message = None;
+        let latest_system_message = None;
+
+        Mailbox {
+            user_tx,
+            user_rx,
+            system_rx,
+            latest_user_message,
+            latest_system_message,
+            state,
+        }
     }
 
     /// Forced receive message from user queue
-    pub async fn recv(&self) -> Envelope<T> {
-        self.inner
-            .user_rx()
+    pub async fn recv(&mut self) -> Envelope<T> {
+        if let Some(old_message) = &self.latest_user_message {
+            old_message.ack().await;
+        }
```

Comment:
Wouldn't it happen too late?

Reply:
I think, you're right. These lines are excessive.